File: java/src/test/java/io/valkey/javabenchmark/engine/RateLimiterTest.java
/*
 * Copyright 2025 the original author or authors.
 */
package io.valkey.javabenchmark.engine;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Unit tests for the {@link RateLimiter} class.
 * 
 * <p>RateLimiter uses a leaky bucket algorithm that enforces a constant rate
 * without burst. Operations are evenly spaced at the configured rate.</p>
 */
class RateLimiterTest {

    @Test
    void createReturnsNullForNegativeRate() {
        RateLimiter limiter = RateLimiter.create(-1);
        assertThat(limiter).isNull();
    }

    @Test
    void createReturnsNullForZeroRate() {
        RateLimiter limiter = RateLimiter.create(0);
        assertThat(limiter).isNull();
    }

    @Test
    void createReturnsInstanceForPositiveRate() {
        RateLimiter limiter = RateLimiter.create(100);
        assertThat(limiter).isNotNull();
    }

    @Test
    void getRatePerSecondReturnsConfiguredRate() {
        RateLimiter limiter = RateLimiter.create(42);
        assertThat(limiter.getRatePerSecond()).isEqualTo(42);
    }

    @Test
    void tryAcquireSucceedsOnFirstCall() {
        RateLimiter limiter = RateLimiter.create(20);
        assertThat(limiter.tryAcquire()).isTrue();
    }

    @Test
    void tryAcquireFailsImmediatelyAfterFirstCall() {
        // With leaky bucket, second call should fail if made immediately
        RateLimiter limiter = RateLimiter.create(20); // 50ms interval
        
        assertThat(limiter.tryAcquire()).isTrue();  // First succeeds
        assertThat(limiter.tryAcquire()).isFalse(); // Second fails (no time elapsed)
    }

    @Test
    void tryAcquireSucceedsAfterInterval() throws Exception {
        int rate = 20; // 50ms interval
        RateLimiter limiter = RateLimiter.create(rate);
        
        assertThat(limiter.tryAcquire()).isTrue();  // First succeeds
        
        // Wait for interval
        Thread.sleep(55); // slightly more than 50ms
        
        assertThat(limiter.tryAcquire()).isTrue();  // Should succeed after interval
    }

    @Test
    void acquireEnforcesConstantRate() throws Exception {
        int rate = 20; // 20 ops/sec = 50ms interval
        RateLimiter limiter = RateLimiter.create(rate);
        
        long startTime = System.currentTimeMillis();
        
        // Acquire 5 operations
        for (int i = 0; i < 5; i++) {
            limiter.acquire();
        }
        
        long durationMs = System.currentTimeMillis() - startTime;
        
        // 5 ops at 20/sec with first immediate: 4 intervals of 50ms = 200ms
        // Allow some tolerance for timing variability
        assertThat(durationMs).isGreaterThanOrEqualTo(180L);
    }

    @Test
    void acquireMaintainsRateOverTime() throws Exception {
        int rate = 20; // 20 ops/sec
        RateLimiter limiter = RateLimiter.create(rate);
        
        long startTime = System.currentTimeMillis();
        int count = 0;
        
        // Run for 1 second
        while (System.currentTimeMillis() - startTime < 1000) {
            limiter.acquire();
            count++;
        }
        
        // Should have executed approximately 20 operations
        // With 5% tolerance: 19-21
        assertThat(count).isBetween(19, 21);
    }

    @Test
    void lowRateCreatesSignificantDelay() throws Exception {
        int rate = 5; // 5 ops/sec = 200ms interval
        RateLimiter limiter = RateLimiter.create(rate);
        
        long startTime = System.currentTimeMillis();
        
        // 3 operations: first immediate, then 2 waits of 200ms
        for (int i = 0; i < 3; i++) {
            limiter.acquire();
        }
        
        long durationMs = System.currentTimeMillis() - startTime;
        
        // Should take ~400ms (2 intervals of 200ms)
        assertThat(durationMs).isGreaterThanOrEqualTo(380L);
    }
}

File: ruby/test/unit/rate_limiter_test.rb
# frozen_string_literal: true

require_relative "../test_helper"

class RateLimiterTest < Minitest::Test
  def test_create_returns_nil_for_negative_rate
    limiter = RespBench::Engine::RateLimiter.create(-1)
    assert_nil limiter
  end

  def test_create_returns_nil_for_zero_rate
    limiter = RespBench::Engine::RateLimiter.create(0)
    assert_nil limiter
  end

  def test_create_returns_instance_for_positive_rate
    limiter = RespBench::Engine::RateLimiter.create(100)
    refute_nil limiter
  end

  def test_rate_per_second_returns_configured_rate
    limiter = RespBench::Engine::RateLimiter.create(42)
    assert_equal 42, limiter.rate_per_second
  end

  def test_try_acquire_succeeds_on_first_call
    limiter = RespBench::Engine::RateLimiter.create(20)
    assert limiter.try_acquire
  end

  def test_try_acquire_fails_immediately_after_first_call
    # With leaky bucket, second call should fail if made immediately
    limiter = RespBench::Engine::RateLimiter.create(20) # 50ms interval

    assert limiter.try_acquire  # First succeeds
    refute limiter.try_acquire  # Second fails (no time elapsed)
  end

  def test_try_acquire_succeeds_after_interval
    rate = 20 # 50ms interval
    limiter = RespBench::Engine::RateLimiter.create(rate)

    assert limiter.try_acquire  # First succeeds

    # Wait for interval
    sleep(0.055) # slightly more than 50ms

    assert limiter.try_acquire  # Should succeed after interval
  end

  def test_acquire_enforces_constant_rate
    rate = 20 # 20 ops/sec = 50ms interval
    limiter = RespBench::Engine::RateLimiter.create(rate)

    start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)

    # Acquire 5 operations
    5.times { limiter.acquire }

    duration_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time) * 1000

    # 5 ops at 20/sec with first immediate: 4 intervals of 50ms = 200ms
    # Allow some tolerance for timing variability
    assert duration_ms >= 180, "Expected duration >= 180ms, got #{duration_ms}ms"
  end

  def test_acquire_maintains_rate_over_time
    rate = 20 # 20 ops/sec
    limiter = RespBench::Engine::RateLimiter.create(rate)

    start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    count = 0

    # Run for 1 second
    while (Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time) < 1.0
      limiter.acquire
      count += 1
    end

    # Should have executed approximately 20 operations
    # With 5% tolerance: 19-21 (matching Java's isBetween(19, 21))
    assert count >= 19 && count <= 21, "Expected 19-21 ops, got #{count}"
  end

  def test_low_rate_creates_significant_delay
    rate = 5 # 5 ops/sec = 200ms interval
    limiter = RespBench::Engine::RateLimiter.create(rate)

    start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)

    # 3 operations: first immediate, then 2 waits of 200ms
    3.times { limiter.acquire }

    duration_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time) * 1000

    # Should take ~400ms (2 intervals of 200ms)
    assert duration_ms >= 380, "Expected duration >= 380ms, got #{duration_ms}ms"
  end
end
